### PR TITLE
GitBlame: RuntimeError when analyzing branch name that is within whitelisted format #508

### LINKED ISSUE #508: GitBlame: RuntimeError when analyzing branch name that is within whitelisted format
```
Exception in thread "main" java.lang.RuntimeException: Error returned from command git blame -w --line-porcelain ".git/logs/refs/heads/haha.txt"on path repos/testrepo-Charlie/testrepo-Charlie :
fatal: no such path '.git/logs/refs/heads/haha.txt' in HEAD

    at reposense.system.CommandRunner.runCommand(CommandRunner.java:55)
    at reposense.git.GitBlame.blame(GitBlame.java:32)
    at reposense.authorship.FileInfoAnalyzer.getGitBlameResult(FileInfoAnalyzer.java:107)
    at reposense.authorship.FileInfoAnalyzer.aggregateBlameAuthorInfo(FileInfoAnalyzer.java:82)
    at reposense.authorship.FileInfoAnalyzer.analyzeFile(FileInfoAnalyzer.java:48)
    at reposense.authorship.AuthorshipReporter.lambda$generateAuthorshipSummary$0(AuthorshipReporter.java:24)
    at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
    at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1492)
    at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
    at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
    at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
    at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
    at reposense.authorship.AuthorshipReporter.generateAuthorshipSummary(AuthorshipReporter.java:26)
    at reposense.report.ReportGenerator.generateReposReport(ReportGenerator.java:78)
    at reposense.RepoSense.main(RepoSense.java:52)
```

To replicate, create a branch and name it with a file extension being analyzed (e.g. `xxx.txt`), and in that branch create a file of the same name, and run the analysis.

### PR DESCRIPTION
**Pull request by emer7**
Fix #508 

```
When generating authorship report, RepoSense checks all the files and directories
in the Repository by listing all of those from the Repository root.

However it also includes .git directory that contains all the git repository files, and
RepoSense will throw RuntimeError if it encountered a branch name that matches the
format whitelist.

Let's fix this issue by excluding .git directory from authorship generation.
```

### CODE REVIEW

**Comment by yamidark** [comment]
@emer7 Also, you can request review from us at the RHS of the PR page, at the reviewers section (on top of the labels), since you are already part of the reposense/devs team

File: src/main/java/reposense/authorship/FileInfoExtractor.java [hunk 1]
```diff
@@ -165,7 +166,7 @@ private static void getAllFileInfo(RepoConfiguration config, Path directory, Lis
         try (Stream<Path> pathStream = Files.list(directory)) {
             for (Path filePath : pathStream.collect(Collectors.toList())) {
                 String relativePath = filePath.toString().substring(config.getRepoRoot().length());
-                if (Files.isDirectory(filePath)) {
+                if (Files.isDirectory(filePath) && !relativePath.equals(GIT_DIRECTORY)) {
                     getAllFileInfo(config, filePath, fileInfos);
```
Comment (yong24s) [inline src/main/java/reposense/authorship/FileInfoExtractor.java]: It is necessary to traverse into .git directory?

The code here does not really follow the let's directive. `Let's fix this issue by excluding .git directory from authorship generation.`

Comment (emer7) [reply src/main/java/reposense/authorship/FileInfoExtractor.java]: .git directory is included in `Files.list(directory)`, so I guess we cannot avoid not traversing it?

Comment (yong24s) [reply src/main/java/reposense/authorship/FileInfoExtractor.java]: I believe skipping it entirely is more representative to the let's directive and this approach is more robust.

```
if (!relativePath.equals(GIT_DIRECTORY)) {
    continue;
}
```

Comment (yamidark) [reply src/main/java/reposense/authorship/FileInfoExtractor.java]: >I believe skipping it entirely is more representative to the let's directive and this approach is more robust.

There is still a need to check if its a directory as well (maybe someone wants to create a `.git` file for some reason)

So should be:
```
if (Files.isDirectory(filePath) && relativePath.equals(GIT_DIRECTORY)) {
    continue;
}
```
instead?

File: src/test/java/reposense/authorship/FileInfoExtractorTest.java [hunk 2]
```diff
@@ -84,6 +85,14 @@ public void extractFileInfos_directoryWithValidWhitelistedName_success() {
         Assert.assertTrue(isFileExistence(Paths.get(".gradle/anything.txt"), files));
     }
 
+    @Test
+    public void extractFileInfos_branchWithValidWhitelistedFormatName_success() {
+        GitCheckout.checkout(config.getRepoRoot(), BRANCH_WITH_VALID_WHITELISTED_FORMAT_NAME_BRANCH);
```
Comment (eugenepeh) [inline src/test/java/reposense/authorship/FileInfoExtractorTest.java]: Just would like to better manage the branch naming. 
Can you change the branch name to this format -> `PR_NUM-CLASS_NAME-TEST_NAME`?
So in this case would be `535-FileInfoExtractorTest-branchWithValidWhitelistedFormatName`. You can put the `.txt` behind if its necessary.

Maybe `FileName` is more appropriate than `FormatName` here. Add a short description if you have to.

File: src/test/java/reposense/authorship/FileInfoExtractorTest.java [hunk 3]
```diff
@@ -25,6 +25,7 @@
     private static final String WINDOWS_ILLEGAL_FILE_NAME_BRANCH = "windows-illegal-filename";
     private static final String EDITED_FILE_INFO_BRANCH = "getEditedFileInfos-test";
     private static final String DIRECTORY_WITH_VALID_WHITELISTED_NAME_BRANCH = "directory-with-valid-whitelisted-name";
+    private static final String BRANCH_WITH_VALID_WHITELISTED_FILE_NAME_BRANCH = "535-FileInfoExtractorTest-branchWithValidWhitelistedFileName.txt";
```
Comment (yamidark) [inline src/test/java/reposense/authorship/FileInfoExtractorTest.java]: This is causing checkstyle to fail in Travis and Appveyor as the line is greater than 120 characters.
Make a line break after the `=`

**Comment by yamidark** [comment]
Slights nits to proposed message:
```
When generating authorship report, RepoSense checks all the files
and directories in the repository by listing all of those from the
repository root.

However, it also includes the .git directory that contains all of the
repository's metadata files, and RepoSense will throw RuntimeError
if it encounters a branch name that matches any of the file format
whitelist.

Let's fix this issue by excluding .git directory from file analysis.
```

**Pull request merged**